lib/src/authentication/token.dart

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import '../source/hosted.dart';
+
+/// Token is a structure for storing authentication credentials for third-party
+/// pub registries. A token holds registry [url], credential [kind] and [token]
+/// itself.
+///
+/// Token could be serialized into and from JSON format structured like
+/// this:
+///
+/// ```json
+/// {
+///   "url": "https://example.com/",
+///   "credential": {
+///     "kind": "Bearer",
+///     "token": "gjrjo7Tm2F0u64cTsECDq4jBNZYhco"
+///   }
+/// }
```

Comment:
As a general guideline it's best to [avoid unnecessary nesting](https://google.github.io/styleguide/jsoncstyleguide.xml?showone=Flattened_data_vs_Structured_Hierarchy#Flattened_data_vs_Structured_Hierarchy).

Perhaps, we should just do:
```js
{
  "url": "https://example.com",
  "token": "<secret-token>"
},
```

If we want to introduce additional authentication mechanisms in the future, we can add them like:
```js
{
  "url": "https://example.com",
  "environmentToken": "NAME_OF_ENV_VAR_CONTAINING_TOKEN"
},
// Or maybe:
{
  "url": "https://example.com",
  "serviceAccountName": "<email of service account>"
},
```
( don't think we'll do service accounts, but if we ever decide to try something like that we could do this way).

Reply:
Yes nesting is unnecessary. But I think we might still need to store kind or something else for storing token type otherwise what we'll to if the token is like that:

```
{
  "url": "https://example.com",
  "token": "<secret-token>",
  "environmentToken": "NAME_OF_ENV_VAR"
}
```

Which one we'll going to use in this case?

Reply:
If we need we could add it in future anyways. So removing it for now..